feat!: improve `MemoryStore`

nfriedly opened the pull request:
I've been thinking about how to make the MemoryStore have per-user reset times and still only a single global timer (like `precise-memory-rate-limit`), but *not* have to iterate through all known clients to remove expired ones (unlike `precise-memory-rate-limit`).

I came up with the idea for a previous and current set of clients, where we could pull active ones forward from previous to current, and then at the end of the time window, just blow away the previous, move the current to previous, and create a new current set.

I then refined it a bit to perform less allocations (and by extension, need less garbage collection) by merely swapping previous and current, and then moving the leftover clients from the old previous to a client pool. With an appropriately sized pool, it should be able to completely avoid allocations for most requests.

I had to tweak the tests to use a positive number for `windowMs`, but after that they all passed.

I've also been thinking about a 'ClusterMemoryStore' where there would be a MemoryStore instance on the primary process, and then each worker would communicate with it via IPC to increment and decrement keys.  Then you would only need an external data store for multi-server deployments. Probably should go in it's own module, though.

nfriedly commented:
(Sorry I've been kind of all over the place recently. I'll try and get a few of these things wrapped up sometime soon...)

nfriedly commented on source/memory-store.ts:
```diff
@@ -60,17 +73,25 @@ export default class MemoryStore implements Store {
+		// Initialize the spare client pool
+		this.pool = []
+
+		// Indicates that init was called more than once.
+		// Could happen if a store was shared between multiple instances.
+		if (this.interval) {
+			clearTimeout(this.interval)
+		}
```
This fixes #326

gamemaker1 commented:
Wow! This looks really promising, I'll try reviewing it in detail soon.

nfriedly commented on source/memory-store.ts:
```diff
@@ -30,16 +20,25 @@ export default class MemoryStore implements Store {
+	/**
+	 * Maximum number of unused clients to keep in the pool
+	 */
+	poolSize = 100
```
Maybe it should take [`os.freemem()`](https://nodejs.org/api/os.html#osfreemem) into account also, since it's essentially trading RAM usage for performance.

nfriedly replied:
Actually, inspecting with chrome dev tools shows that a single client object is only ~152 bytes, so 60k would be less than 1mb. So `os.freemem()` won't really have much affect.

I put together a quick test https://gist.github.com/nfriedly/fdba25bcf9f23d18fe8ac5bd1bad5dac and 5000 seems to be the winner from a few runs, but it honestly doesn't make much difference either way.

gamemaker1 commented on source/memory-store.ts:
```diff
@@ -30,16 +24,33 @@ export default class MemoryStore implements Store {
+	 *
+	 * Improves performance, at the expense of a small amount of RAM.
+	 *
+	 * Each individual entry takes up 152 bytes.
+	 * In one benchmark, the total time taken to handle 100M requests was reduced from 70.184s to 47.862s (~32% improvement) with ~5.022 MB extra RAM used.
```
Should we add this benchmark to the repo as `test/benchmark/memory-store.benchmark.ts`?

nfriedly replied:
No, at least not in its current state. It's pretty ad hoc, and coupled to the implementation. I was making changes to the testing script as I changed the implementation details, so that's not even the current iteration.

Also it's very dependent on the machine you run it on, so I don't think it would be very useful to compare across machines.

gamemaker1 replied:
Oh alright then

gamemaker1 commented on source/memory-store.ts:
```diff
@@ -102,9 +117,9 @@ export default class MemoryStore implements Store {
+		const client = this.getClient(key)
 
-		if (current) this.hits[key] = current - 1
+		if (client.totalHits > 1) client.totalHits--
```
Shouldn't this be `> 0`?

nfriedly replied:
No, because we don't want it to ever go negative.

gamemaker1 replied:
But if the number of hits is 1 it should be able to decrement it to 0 right?

nfriedly replied:
🤦 yes.

gamemaker1 commented on source/memory-store.ts:
```diff
@@ -102,9 +117,9 @@ export default class MemoryStore implements Store {
+		const client = this.getClient(key)
 
-		if (current) this.hits[key] = current - 1
+		if (client.totalHits > 1) client.totalHits--
```
Shouldn't this be `> 0`?

gamemaker1 commented on source/memory-store.ts:
```diff
@@ -136,5 +152,72 @@ export default class MemoryStore implements Store {
+		// Fill up the pool with expired clients
+		for (const client of this.previous.values()) {
+			if (poolSpace > 0) {
+				this.pool.push(client)
```
Should we run `resetClient` on the client when adding it to the pool? That way, the pool might take up less resources.

nfriedly replied:
Nah, it takes up the same amount of resources either way, and we have to reset the date when we pull it out of the pool, so it would just be doing extra work to also reset it when we put it in.

gamemaker1 replied:
Ah I didn't consider the date part, makes sense not to do it here then.

nfriedly commented on source/memory-store.ts:
```diff
@@ -155,69 +167,96 @@ export default class MemoryStore implements Store {
 			client = this.previous.get(key)!
 		} else if (this.pool.length > 0) {
+			// If it's in neither the `current` nor the `previous` maps, animate a corpse
+			// from the pool. Spoooooooooooky!
```
👻 😂

nfriedly commented on source/memory-store.ts:
```diff
@@ -30,19 +36,37 @@ export default class MemoryStore implements Store {
+	 * Used to calculate how many entries to keep in the pool.
 	 */
-	resetTime!: Date
+	numberOfCreatedClients = 0
+	recentClientsCount: number[] = []
```
Hum.. I concede that the old names weren't great, but this doesn't feel right either.

The first one is the number of keys that we've seen during the current window that we didn't see during the previous window. Ideally pool should begin the window with exactly this many Client objects, so that 0 are created.

The second one is a series of previous values for the first. So it should probably just be the plural form of the name of the first one. 

Some ideas: 
* `idealPoolSize` & `idealPoolSizes`?
* `currentNewKeys` & `previousNewKeys`
* `uncachedCount` & `uncachedCounts`

I don't know. Naming things is hard.

gamemaker1 replied:
`uncachedCount` & `uncachedCounts` sounds good 👍🏾

nfriedly replied:
done in d89e514

nfriedly commented on source/memory-store.ts:
```diff
@@ -155,69 +167,96 @@ export default class MemoryStore implements Store {
 		let client
 		if (this.previous.has(key)) {
+			// If it's in the `previous` map, return it and bump it up to the `current` map.
 			client = this.previous.get(key)!
```
Oh, I think there's a bug here. Client should be *removed* from `previous`, this just gets a reference to it. The client could potentially end up in the `pool` and `current` at the same time, and then get assigned to a new key, thus ending up in `current` more than once and combining the hits from both keys.

I should write a test to make that scenario happen, and then fix the code.

nfriedly replied:
Fixed in e0801ce

nfriedly commented:
Ugh, I need to redo the performance testing, that bug invalidated everything I did.

nfriedly commented:
Ok, now that I fixed the bug with not deleting clients from previous when moving to current, I re-tested and found that the `pool` actually makes things slightly slower. So I removed it.

I also updated the Use Cases section of the readme although, honestly, we could probably delete it at this point. I wrote that because of how often people were confused by the behavior of the old MemoryStore, and I think the new one will just do what they expect.

nfriedly commented:
Ok, I believe this is now ready to go. 

It's simplified a lot, because my earlier implementation was too clever by half.

gamemaker1 reviewed (approved):
LGTM!

> Ok, now that I fixed the bug with not deleting clients from previous when moving to current, I re-tested and found that the pool actually makes things slightly slower. So I removed it.

Looks like we didn't need to play with ghosts after all XD
